File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/DecodeString.java
package com.only4.algorithm.version4j.leetcode;

import java.util.ArrayDeque;

public class DecodeString {

    static class RepeatItem {
        StringBuilder content = new StringBuilder();
        int count;

        public RepeatItem(int count) {
            this.count = count;
        }

        @Override
        public String toString() {
            return content.toString().repeat(count);
        }
    }

    public String decodeString(String s) {
        ArrayDeque<RepeatItem> stack = new ArrayDeque<>();
        stack.add(new RepeatItem(1));
        StringBuilder number = new StringBuilder();

        for (char c : s.toCharArray()) {
            if (Character.isDigit(c)) {
                number.append(c);
            } else if (c == '[') {
                int count = Integer.parseInt(number.toString());
                number = new StringBuilder();
                stack.addLast(new RepeatItem(count));
            } else if (c == ']') {
                RepeatItem item = stack.removeLast();
                stack.peekLast().content.append(item.toString());
            } else {
                stack.peekLast().content.append(c);
            }
        }

        return stack.peekLast().toString();
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/decodeString.kt
package com.only4.algorithm.version4k.leetcode

class RepeatItem(val count: Int) {
    val content = StringBuilder()

    override fun toString(): String = content.toString().repeat(count)
}

fun decodeString(s: String): String {
    val stack = ArrayDeque<RepeatItem>().apply { add(RepeatItem(1)) }
    var number = StringBuilder()

    for (c in s) {
        when (c) {
            in '0'..'9' -> number.append(c)
            '[' -> {
                val count = number.toString().toInt()
                stack.add(RepeatItem(count))
                number = StringBuilder()
            }
            ']' -> {
                val item = stack.removeLast()
                stack.last().content.append(item.toString())
            }

            else -> stack.last().content.append(c)
        }
    }

    return stack.last().toString()
}
